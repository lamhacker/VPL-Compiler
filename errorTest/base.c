#include "base.h"

float* mallocVec(int num)
{
    float* vec = malloc(sizeof(float) * num + SSE_ALIGN);
	vec = (float *) align(vec, SSE_ALIGN);
	return vec;
}

void freeVec(float* vec)
{
    free(vec);
}

void setValue(float* vec, float val, int num)
{
    for (int i = 0; i < num; i++)
    {
        vec[i] = val;
    }
}

void printVec(char* rep, float* vec, int num)
{
	printf("%s: ", rep);
    for (int i = 0; i < num; i++)
    {   
        printf("%f ", vec[i]);
    }
	printf("\n");
}


