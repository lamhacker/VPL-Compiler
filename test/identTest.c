#include "base.h"
#define NUM (4)

extern void ident1(long, float*, float*);
extern void ident2(long, float*, float*, float*, float*, float*);

int main(void) {
	
	float* a = mallocVec(NUM);
	float* b = mallocVec(NUM);
	float* c = mallocVec(NUM);
	float* d = mallocVec(NUM);
	float* e = mallocVec(NUM);
	
	printf("ident1: \n");
    setValue(a, 1, NUM);
	setValue(b, 2, NUM);
    ident1(NUM, a, b);
	printVec("a", a, NUM);
	printVec("b", b, NUM);
	
	printf("ident2: \n");
    setValue(a, 1, NUM);
	setValue(b, 2, NUM);
	setValue(c, 3, NUM);
	setValue(d, 4, NUM);
	setValue(e, 5, NUM);
	ident2(NUM, a, b, c, d, e);
	printVec("a", a, NUM);
	printVec("b", b, NUM);
	printVec("c", c, NUM);
	printVec("d", d, NUM);
	printVec("e", e, NUM);
	
	freeVec(a);
	freeVec(b);
	freeVec(c);
	freeVec(d);
	freeVec(e);
	
	return 0;
}
