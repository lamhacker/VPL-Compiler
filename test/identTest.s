
.text
.global ident1
.type ident1, @function
.p2align 4,,15

ident1:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    

    # Condition
	
    # <load source address into %rax> 
    
    # place address of parameter into %rax
    movq %rsi, %rax


    xorps %xmm1, %xmm1
    movq %rdi, %rbx
    shrq $2, %rbx

    jz .loop_end1

.loop_begin1:

    addps (%rax), %xmm1 
    
    # exist when %rax is not pointing to constant
    addq $16, %rax

    decq %rbx
    jnz .loop_begin1

.loop_end1:
    xorps %xmm0, %xmm0
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0

    ucomiss .L0.0, %xmm0
    jb .falsebranch0
    jmp .truebranch0


    .align 4
.L0.0:
    .float 0.0



.truebranch0:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rdx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end2

.loop_begin2:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin2
.loop_end2:


	jmp .endif0

.falsebranch0:

.endif0:


    popq    %rbx
    leave
    ret

.text
.global ident2
.type ident2, @function
.p2align 4,,15

ident2:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    

    # Condition
	
    # <load source address into %rax> 
    
    # place address of parameter into %rax
    movq %rsi, %rax


    xorps %xmm1, %xmm1
    movq %rdi, %rbx
    shrq $2, %rbx

    jz .loop_end4

.loop_begin4:

    addps (%rax), %xmm1 
    
    # exist when %rax is not pointing to constant
    addq $16, %rax

    decq %rbx
    jnz .loop_begin4

.loop_end4:
    xorps %xmm0, %xmm0
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0

    ucomiss .L100.0, %xmm0
    jb .truebranch3
    jmp .falsebranch3




.truebranch3:


    # Condition
	
    # <load source address into %rax> 
    
    # place address of parameter into %rax
    movq %rdx, %rax


    xorps %xmm1, %xmm1
    movq %rdi, %rbx
    shrq $2, %rbx

    jz .loop_end6

.loop_begin6:

    addps (%rax), %xmm1 
    
    # exist when %rax is not pointing to constant
    addq $16, %rax

    decq %rbx
    jnz .loop_begin6

.loop_end6:
    xorps %xmm0, %xmm0
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0

    ucomiss .L100.0, %xmm0
    jb .truebranch5
    jmp .falsebranch5




.truebranch5:


    # Condition
	
    # <load source address into %rax> 
    
    # place address of parameter into %rax
    movq %rcx, %rax


    xorps %xmm1, %xmm1
    movq %rdi, %rbx
    shrq $2, %rbx

    jz .loop_end8

.loop_begin8:

    addps (%rax), %xmm1 
    
    # exist when %rax is not pointing to constant
    addq $16, %rax

    decq %rbx
    jnz .loop_begin8

.loop_end8:
    xorps %xmm0, %xmm0
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0

    ucomiss .L100.0, %xmm0
    jb .truebranch7
    jmp .falsebranch7




.truebranch7:


    # Condition
	
    # <load source address into %rax> 
    
    # place address of parameter into %rax
    movq %r8, %rax


    xorps %xmm1, %xmm1
    movq %rdi, %rbx
    shrq $2, %rbx

    jz .loop_end10

.loop_begin10:

    addps (%rax), %xmm1 
    
    # exist when %rax is not pointing to constant
    addq $16, %rax

    decq %rbx
    jnz .loop_begin10

.loop_end10:
    xorps %xmm0, %xmm0
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0
    shufps $147,%xmm1,%xmm1
    addss %xmm1,%xmm0

    ucomiss .L100.0, %xmm0
    jb .truebranch9
    jmp .falsebranch9


    .align 4
.L100.0:
    .float 100.0



.truebranch9:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %r8, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %r9, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end11

.loop_begin11:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin11
.loop_end11:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rcx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %r8, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end12

.loop_begin12:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin12
.loop_end12:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rdx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rcx, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end13

.loop_begin13:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin13
.loop_end13:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rsi, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rdx, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end14

.loop_begin14:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin14
.loop_end14:

    # load source address into %rax
    
    # place address of 0.0 into %rax
    movq $.const0.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end15

.loop_begin15:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin15
.loop_end15:


	jmp .endif9

.falsebranch9:

.endif9:


	jmp .endif7

.falsebranch7:

.endif7:


	jmp .endif5

.falsebranch5:

.endif5:


	jmp .endif3

.falsebranch3:

.endif3:


    popq    %rbx
    leave
    ret

.data 
.align 16 
.const0.0:
    .float 0.0 
    .float 0.0 
    .float 0.0 
    .float 0.0

