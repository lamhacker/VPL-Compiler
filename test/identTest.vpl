func ident1(a, b)
	if a >= 0 then
		a = b;
	endif
end

func ident2(a, b, c, d, e)
	if a < 100 then
		if b < 100 then
			if c < 100 then
				if d < 100 then
					e = d;
					d = c;
					c = b;
					b = a;
					a = 0;
				endif
			endif
		endif
	endif
end
