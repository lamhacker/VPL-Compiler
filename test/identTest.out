ident1: 
a: 2.000000 2.000000 2.000000 2.000000 
b: 2.000000 2.000000 2.000000 2.000000 
ident2: 
a: 0.000000 0.000000 0.000000 0.000000 
b: 1.000000 1.000000 1.000000 1.000000 
c: 2.000000 2.000000 2.000000 2.000000 
d: 3.000000 3.000000 3.000000 3.000000 
e: 4.000000 4.000000 4.000000 4.000000 
